### PR TITLE
Fix our busted CFKeyedArchiverUID support. Fixes #649.

### LINKED ISSUE #649: Binary property lists and keyed archives from the reference platform may behave badly
Due to the lack of a generic umbrella bridged type, `CFKeyedArchiverUID`s, which is used in keyed archives from iOS and OS X, will not be deserialized properly. This will result in a crash when you attempt to use `NSKeyedUnarchiver` or `NSPropertyListSerialization` with archives created on the reference platform.

Normal property lists from iOS and OS X are not affected, as they will not contain UID objects.

XML archives are not affected, as they encode UIDs via `NSDictionary` (as `<dict><key>CF$UID</key><integer>1</integer></dict>`).

Reproduction case:

``` objc
// The only way to represent a raw CFKeyedArchiverUID is via binary property list.
// This property list contains:
// 0x00:        u8[8] header                       (bplist00)
// -- object table --
// 0x08: UID (0x80 0xFF)
// -- offset table --
// 0x0A: 0x08
// -- plist trailer --
// 0x0B - 0x0F: u8[5] unused                       (0x00 * 5)
// 0x10:        u8    sort version                 (0x00)
// 0x11:        u8    size of offset table entries (0x01)
// 0x12:        u8    size of embedded objects     (0x01)
// 0x13 - 0x1A: u64   number of objects            (0x01)
// 0x1B - 0x22: u64   toplevel object              (0x00)
// 0x23 - 0x2A: u64   offset to offset table       (0x0A)

const char rawPlist[] = {
    0x62, 0x70, 0x6c, 0x69, 0x73, 0x74, 0x30, 0x30, 0x80, 0xff, 0x08, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0a
};
NSData* plistData = [NSData dataWithBytes:rawPlist length:_countof(rawPlist)];
NSObject* UID = [NSPropertyListSerialization propertyListFromData:plistData mutabilityOption:0 format:nullptr errorDescription:nullptr];
```

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
For some reason, we had a separate binary plist writer too. It was only being used for NSDictionary and NSArray `writeToFile:`.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1633)

### CODE REVIEW

File: Frameworks/Foundation/NSKeyedArchiver.mm [hunk 1]
```diff
@@ -65,12 +66,7 @@ static inline void _checkKey(NSString* aKey, NSMutableDictionary* _enc) {
 * Make a dictionary referring to the object at ref in the array of all objects.
 */
 static id makeReference(unsigned ref) {
```
Comment (rajsesh-msft) [inline Frameworks/Foundation/NSKeyedArchiver.mm]: >ref [](start = 33, length = 3)

nit: ref is a special word in some languages and can confuse some of us, could we rename this?

**Comment by bbowman** [comment]
:shipit:

**Comment by rajsesh-msft** [comment]
:shipit:

**Comment by rajsesh-msft** [comment]
It would be a good idea to test islandwoodnews with this change.

**Comment by msft-Jeyaram** [comment]
:shipit:

**Comment by DHowett-MSFT** [comment]
@rajsesh-msft IWNews launches and loads its config properly. It doesn't look like it archives (via keyed archival) any data at runtime, however!

**Pull request merged**